src/com/twitter/intellij/pants/settings/PantsProjectSettingsControl.java

```diff
@@ -35,103 +36,121 @@
 
-    myLibsWithSourcesCheckBox = new JBCheckBox(PantsBundle.message("pants.settings.text.with.sources.and.docs"));
-    myLibsWithSourcesCheckBox.setSelected(false);
-    content.add(myLibsWithSourcesCheckBox, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
+    content.add(myLibsWithSourcesCheckBox, lineConstraints);
+    content.add(myEnableIncrementalImportCheckBox, lineConstraints);
+    content.add(myUseIdeaProjectJdkCheckBox, lineConstraints);
 
-    myEnableIncrementalImport = new JBCheckBox(PantsBundle.message("pants.settings.text.with.incremental.import"));
-    myEnableIncrementalImport.setSelected(false);
-    content.add(myEnableIncrementalImport, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
+    mySettings.getTargetSpecs().forEach(spec -> myTargetSpecsBox.addItem(spec, spec, true));
 
-    final JLabel targetsLabel = new JBLabel(PantsBundle.message("pants.settings.text.targets"));
-    content.add(targetsLabel, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
+    final JBLabel targetsLabel = new JBLabel(PantsBundle.message("pants.settings.text.targets"));
+    content.add(targetsLabel, lineConstraints);
 
-    myTargetSpecs = new CheckBoxList<String>();
-    content.add(ScrollPaneFactory.createScrollPane(myTargetSpecs), ExternalSystemUiUtil.getFillLineConstraints(0));
+    JScrollPane pane = ScrollPaneFactory.createScrollPane(myTargetSpecsBox);
+    content.add(pane, lineConstraints);
+  }
+
+  // It is silly `CheckBoxList` does not provide an iterator.
+  private List<String> getSelectedTargetSpecs() {
+    List<String> selectedSpecs = new ArrayList<>();
+    for (int i = 0; i < myTargetSpecsBox.getModel().getSize(); i++) {
+      JCheckBox checkBox = (JCheckBox) myTargetSpecsBox.getModel().getElementAt(i);
+      if (checkBox.isSelected()) {
+        selectedSpecs.add(checkBox.getText());
+      }
+    }
+    return selectedSpecs;
   }
 
   @Override
   protected boolean isExtraSettingModified() {
-    return !myTargetSpecs.getSelectedValuesList().isEmpty();
+
+    PantsProjectSettings newSettings = new PantsProjectSettings(
+      getSelectedTargetSpecs(),
+      // Project path is not visible to user, so it will stay the same.
+      getInitialSettings().getExternalProjectPath(),
+      myLibsWithSourcesCheckBox.isSelected(),
+      myEnableIncrementalImportCheckBox.isSelected(),
+      myUseIdeaProjectJdkCheckBox.isSelected()
+    );
+
+    return !newSettings.equals(getInitialSettings());
   }
 
   @Override
   protected void resetExtraSettings(boolean isDefaultModuleCreation) {
-    final String externalProjectPath = getInitialSettings().getExternalProjectPath();
-    if (!StringUtil.isEmpty(externalProjectPath)) {
-      onProjectPathChanged(externalProjectPath);
-    }
   }
 
   public void onProjectPathChanged(@NotNull final String projectPath) {
-    myTargetSpecs.clear();
+    if (lastPath.equals(projectPath)) {
+      lastPath = projectPath;
```

Comment:
Why assign here if they are equal? This could be reduced to being just a guard clause.